pallets/phala/src/fat.rs

```diff
@@ -0,0 +1,284 @@
+		InvalidSender,
+		WorkerNotFound,
+	}
+
+	type CodeHash<T> = <T as frame_system::Config>::Hash;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		T: crate::mq::Config + crate::registry::Config,
+	{
+		#[pallet::weight(0)]
+		pub fn upload_code(origin: OriginFor<T>, code: Vec<u8>) -> DispatchResult {
+			ensure_signed(origin)?;
+			let code_hash = T::Hashing::hash(&code);
+			ContractCode::<T>::insert(&code_hash, &code);
+			Self::deposit_event(Event::CodeUploaded(code_hash));
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn upload_code_to_cluster(
+			origin: OriginFor<T>,
+			code: Vec<u8>,
+			cluster_id: ContractClusterId,
+		) -> DispatchResult {
+			let origin: T::AccountId = ensure_signed(origin)?;
+			// TODO.shelven: check permission?
+			Self::push_message(ContractOperation::UploadCodeToCluster {
+				origin,
+				code,
+				cluster_id,
+			});
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn instantiate_contract(
+			origin: OriginFor<T>,
+			// #[pallet::compact] endowment: BalanceOf<T>,
+			// #[pallet::compact] gas_limit: Weight,
+			code_index: CodeIndex<CodeHash<T>>,
+			data: Vec<u8>,
+			salt: Vec<u8>,
+			deploy_to: DeployTarget,
+		) -> DispatchResult {
+			let deployer = ensure_signed(origin)?;
+
+			match code_index {
+				CodeIndex::NativeCode(_) => {}
+				CodeIndex::WasmCode(code_hash) => {
+					ensure!(
+						ContractCode::<T>::contains_key(code_hash),
+						Error::<T>::CodeNotFound
+					);
+				}
+			}
+
+			let (cluster_id, deploy_workers) = match deploy_to {
+				DeployTarget::Cluster(cluster_id) => {
+					ensure!(
+						ClusterWorkers::<T>::contains_key(cluster_id),
+						Error::<T>::ContractClusterNotFound
+					);
+					let workers = ClusterWorkers::<T>::get(cluster_id).expect("checked; qed.");
+					(cluster_id, workers)
+				}
+				DeployTarget::NewGroup(deploy_workers) => {
+					ensure!(deploy_workers.len() > 0, Error::<T>::NoWorkerSpecified);
+
+					let counter = ContractClusterCounter::<T>::mutate(|counter| {
+						*counter += 1;
+						*counter
+					});
+					let cluster_id = ContractClusterId::from_low_u64_be(counter);
+					(cluster_id, deploy_workers)
+				}
+			};
+
+			let mut workers = Vec::new();
+			for worker in deploy_workers.clone().into_iter() {
+				let worker_info =
+					registry::Workers::<T>::try_get(&worker).or(Err(Error::<T>::WorkerNotFound))?;
+				workers.push(WorkerIdentity {
+					pubkey: worker_info.pubkey,
+					ecdh_pubkey: worker_info.ecdh_pubkey,
+				});
+			}
+			if !ClusterWorkers::<T>::contains_key(cluster_id) {
+				ClusterWorkers::<T>::insert(&cluster_id, deploy_workers);
+			}
+
+			// keep syncing with get_contract_id() in crates/phactory/src/contracts/mod.rs
+			fn get_contract_id(
+				deployer: &[u8],
+				code_hash: &[u8],
+				cluster_id: &[u8],
+				salt: &[u8],
+			) -> ContractId {
+				let buf: Vec<_> = deployer
+					.iter()
+					.chain(code_hash)
+					.chain(cluster_id)
+					.chain(salt)
+					.cloned()
+					.collect();
```

Comment:
Looks like a good idea to make a common function `contract_id_preimage(...)` that returns the preimage. Then let the caller to calculate the blake2_256 hash. The difficulty is that different places may use their own blake2s impelmentation

Reply:
I implement the `contract_id()` for `ContractInfo` except for crate/pink/contract.rs, since it defines cluster id to 
low-level `Vec<u8>` install of `ContractClusterId`.